src/interchange/src/avro.rs

```diff
@@ -1151,46 +1189,76 @@ impl DebeziumDeduplicationState {
+            old_max_pos: &'a usize,
+            old_max_row: &'a usize,
+            old_offset: &'a Option<i64>,
+        }
+        let should_skip = match self.binlog_offsets.get_mut(file) {
+            Some((old_max_pos, old_max_row, old_offset)) => {
+                if (*old_max_pos, *old_max_row) >= (pos, row) {
+                    Some(SkipInfo {
+                        old_max_pos,
+                        old_max_row,
+                        old_offset,
+                    })
+                } else {
+                    // update the debezium high water mark
+                    *old_max_pos = pos;
+                    *old_max_row = row;
+                    *old_offset = coord;
+                    None
                 }
-                true
             }
+            None => {
+                // The extra lookup is fine - this is the cold path.
+                self.binlog_offsets
+                    .insert(file.to_owned(), (pos, row, coord));
+                None
+            }
+        };
 
-            DebeziumDeduplicationState::Full { seen_offsets } => {
+        match &mut self.full {
+            TrackFull::No => should_skip.is_none(),
+            TrackFull::SeenOffsets(seen_offsets) => {
                 if let Some(seen_offsets) = seen_offsets.get_mut(file) {
-                    if seen_offsets.insert((pos, row)) {
-                        true
-                    } else {
-                        warn!("Source {}:{} already ingested binlog coordinates ({}, {}) in file {}. Skipping record.",
-                              debug_name, worker_idx, pos, row, file);
-                        false
+                    let is_new = seen_offsets.insert((pos, row));
+
+                    match (is_new, should_skip) {
+                        // new item that correctly does not violate debezium guarantees
+                        (true, None) => {}
+                        // new item that violates debezium guarantees that the no new
+                        // records will ever be sent with a position below the highest
+                        // position ever seen
+                        (true, Some(skipinfo)) => {
+                            // TODO: This would be nicer if it included the wall-clock kafka event time
+                            warn!(
+                                "Source: {}:{} created a new record behind the highest point in binlog_file={} \
+                                 new_record_position={}:{} new_record_kafka_offset={} old_max_position={}:{}",
+                                debug_name, worker_idx, file, pos, row, coord.unwrap_or(-1),
+                                skipinfo.old_max_pos, skipinfo.old_max_row
+                            );
+                        }
+                        // already seen item debezium says it *may* repeat things, so this is fine
+                        (false, Some(skipinfo)) => {
+                            info!(
```

Comment:
Should this be something we don't print by default? Like trace or debug? Will probably cause a lot of spam otherwise...

Reply:
If I wrote this correctly this should print the same amount that we were issuing `warn` statements before, was that too spammy? I'll move this to debug anyway, it conceptually debug level, I just didn't want to alter diagnostics that folks were relying on.